### PR TITLE
User activity

### PR DESCRIPTION
**Pull request by 8masterofpuppets8**
We want to show users what their followees are doing. Using the activity, each time a user does a specific action (comment, create, like, follow), we will notify all their followers via an Activity event which they can request on demand.

### CODE REVIEW

File: src/main/java/com/graffitab/server/service/user/UserService.java [hunk 1]
```diff
@@ -96,6 +101,15 @@
 
 	private static ThreadLocal<User> threadLocalUserCache = new ThreadLocal<>();
 
+	@Transactional
+	public List<Long> getUserFollowersIds(User user) {
+		Query query = userDao.createNamedQuery("User.getFollowerIds");
+		query.setParameter("currentUser", user);
+		@SuppressWarnings("unchecked")
```
Comment (dfernandezm) [inline src/main/java/com/graffitab/server/service/user/UserService.java]: Put this always at method level

File: src/main/java/com/graffitab/server/persistence/model/activity/ActivityComment.java [hunk 2]
```diff
@@ -43,4 +43,19 @@ public ActivityComment(User commenter, Streamable commentedStreamable, Comment c
...
 	}
+
+	@Override
+	public boolean isSameActivity(Activity other) {
+		if (!isSameTypeOfActivity(other)) {
+			return false;
+		}
+
+		ActivityComment activityComment = (ActivityComment) other;
+		return activityComment.getCommentedStreamable().equals(this.commentedStreamable);
```
Comment (dfernandezm) [inline src/main/java/com/graffitab/server/persistence/model/activity/ActivityComment.java]: is `equals()` defined for `Streamable`?

Comment (8masterofpuppets8) [inline src/main/java/com/graffitab/server/persistence/model/activity/ActivityComment.java]: We have it defined using Lombok and the annotation @EqualsAndHashCode.

File: src/main/java/com/graffitab/server/service/paging/ActivityPagingService.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		ListItemsResult<K> listItemsResult = new ListItemsResult<>();
+		listItemsResult.setItems(itemDtos);
+		listItemsResult.setResultsCount(items.getResultsCount());
+		listItemsResult.setMaxResultsCount(items.getMaxResultsCount());
+		listItemsResult.setOffset(items.getOffset());
+
+		return listItemsResult;
+	}
+
+	private List<ActivityContainer> groupActivities(List<Activity> activities, Integer numberOfItemsInGroup) {
```
Comment (dfernandezm) [inline src/main/java/com/graffitab/server/service/paging/ActivityPagingService.java]: Consider using a custom mapper here, from List<Activity> to List<ActivityContainer>

File: src/main/java/com/graffitab/server/service/paging/ActivityPagingService.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+		numberOfItemsInGroup = numberOfItemsInGroup != null ? Math.abs(numberOfItemsInGroup) : NUM_GROUP_ITEMS_DEFAULT_VALUE;
+
+		// Guard against malicious input.
+		if (numberOfItemsInGroup > NUM_GROUP_ITEMS_MAX_VALUE)
+			numberOfItemsInGroup = NUM_GROUP_ITEMS_MAX_VALUE;
+
+		// Get list of entities.
+		PagedList<T> items = getItems(query, offset, count);
+		@SuppressWarnings("unchecked")
+		List<ActivityContainer> groupedActivities = groupActivities((List<Activity>) items, numberOfItemsInGroup);
```
Comment (dfernandezm) [inline src/main/java/com/graffitab/server/service/paging/ActivityPagingService.java]: We could try to do this grouping in the query, but for now I would leave it as it is.

**Pull request closed without merging**